firebase-perf/src/main/java/com/google/firebase/perf/application/FrameMetricsRecorder.java

```diff
@@ -0,0 +1,174 @@
+      return;
+    }
+    frameMetricsAggregator.add(activity);
+    isRecording = true;
+  }
+
+  /**
+   * Stops recording FrameMetrics for the activity window.
+   *
+   * @return FrameMetrics accumulated during the current recording.
+   */
+  public Optional<PerfFrameMetrics> stop() {
+    if (!isRecording) {
+      logger.warn("Cannot stop because no recording was started");
+      return Optional.absent();
+    }
+    if (!subTraceMap.isEmpty()) {
+      logger.warn(
+          "Sub-traces are still ongoing! Sub-traces should be stopped first before stopping Activity screen trace.");
+    }
+    Optional<PerfFrameMetrics> data = this.snapshot();
+    try {
+      frameMetricsAggregator.remove(activity);
+    } catch (IllegalArgumentException err) {
+      logger.debug(
+          "View not hardware accelerated. Unable to collect FrameMetrics. %s", err.toString());
+      return Optional.absent();
+    }
+    frameMetricsAggregator.reset();
+    isRecording = false;
+    return data;
+  }
+
+  /**
+   * Starts a sub-trace in the current recording.
+   *
+   * @param key a UI state to associate this sub-trace with. e.g.) fragment
+   */
+  public void startSubTrace(Fragment key) {
+    if (!isRecording) {
+      logger.warn("Cannot start sub-trace because FrameMetricsAggregator is not recording");
+      return;
+    }
+    if (subTraceMap.containsKey(key)) {
+      logger.warn(
+          "Cannot start sub-trace because one is already ongoing with the key %s",
+          key.getClass().getSimpleName());
+      return;
+    }
+    Optional<PerfFrameMetrics> snapshot = this.snapshot();
+    if (!snapshot.isAvailable()) {
+      logger.warn("startSubTrace(%s): snapshot() failed", key.getClass().getSimpleName());
+      return;
+    }
+    subTraceMap.put(key, snapshot.get());
+  }
+
+  /**
+   * Stops the sub-trace associated with the given UI state.
+   *
+   * @param key the UI state associated with the sub-trace to be stopped.
+   * @return FrameMetrics accumulated during this sub-trace.
+   */
+  public Optional<PerfFrameMetrics> stopSubTrace(Fragment key) {
+    if (!isRecording) {
+      logger.warn("Cannot stop sub-trace because FrameMetricsAggregator is not recording");
+      return Optional.absent();
+    }
+    if (!subTraceMap.containsKey(key)) {
+      logger.warn(
+          "Sub-trace associated with key %s was not started or does not exist",
+          key.getClass().getSimpleName());
+      return Optional.absent();
+    }
+    PerfFrameMetrics snapshotStart = subTraceMap.remove(key);
+    Optional<PerfFrameMetrics> snapshotEnd = this.snapshot();
+    if (!snapshotEnd.isAvailable()) {
+      logger.warn("stopSubTrace(%s): snapshot() failed", key.getClass().getSimpleName());
+      return Optional.absent();
+    }
+    return Optional.of(snapshotEnd.get().subtract(snapshotStart));
+  }
+
+  /**
+   * Calculate total frames, slow frames, and frozen frames from SparseIntArray[] recorded by {@link
```

Comment:
If it is calculate, should we call the method as "calculateMetrics"? Based on what is happening in the function, my recommendation is `snapshotFrameMetrics` and change the comment to `Snapshot frame metrics from Frame Metrics Aggregator`.

Reply:
I forgot to update the comments. This used to have the logic in `FrameMetricsCalculator` copied over, but now it just uses `FrameMetricsCalculator`, so yes I will update the comment so it's about "snapshotting" and not about "calculating".